Refactor config loading

SzymonPobiega opened the pull request:
Clean up the code for loading NH configuration in order to prepare for API changes in future

SzymonPobiega commented:
@Particular/nservicebus please review

danielmarbach commented on src/NServiceBus.NHibernate.Tests/Persistence/ThrowIfRequiredPropertiesAreMissing.cs:
```diff
@@ -10,13 +9,13 @@ public class ThrowIfRequiredPropertiesAreMissing
         [Test]
         public void Should_throw_if_minimum_properties_not_set()
```
The test says `should_throw` but it doesn't

danielmarbach commented on src/NServiceBus.NHibernate.Tests/Persistence/ThrowIfRequiredPropertiesAreMissing.cs:
```diff
@@ -10,13 +9,13 @@ public class ThrowIfRequiredPropertiesAreMissing
         }
 
         [Test]
         public void Should_not_throw_if_minimum_properties_are_set()
```
The test says `should_not_throw` but the assertions is `IsTrue`

danielmarbach commented on src/NServiceBus.NHibernate.Tests/Persistence/ThrowIfRequiredPropertiesAreMissing.cs:
```diff
@@ -25,7 +24,7 @@ public void Should_not_throw_if_minimum_properties_are_set()
         [Test]
         public void Should_not_throw_if_connectionStringName_is_used_instead_of_connectionString()
```
The test says `should_not_throw` but the assertions is `IsTrue`

danielmarbach commented:
@SzymonPobiega can you describe a bit more the purpose of this PR. I see that you changed the behavior. Previously it did throw when configuration properties were missing, now you don't. What is the reasoning behind the behavior change?

danielmarbach commented on src/NServiceBus.NHibernate/Internal/ConfigureNHibernate.cs:
```diff
@@ -20,57 +19,62 @@ namespace NServiceBus.Persistence.NHibernate
+            }
+            if (settingsKeys.Length == 0)
+            {
+                throw new InvalidOperationException("At least one setting key is required");
```
Isn't this exception pretty much useless for the user? You remove the sample in the exception message. In order to have a clear picture what I should do in case of this exception a sample in the message would be helpful

SzymonPobiega replied:
The new exceptions are raised by `ValidateConfigurationViaConfigFile` and `ValidateConfigurationViaCode` methods to give the user some idea of what to do. This is still not ideal because if a user missed adding `NServiceBus/Persistence` connection string, he would be given three most popular config options in order of precedence but 95% of the time he should go with option 2

danielmarbach commented on src/NServiceBus.NHibernate/Internal/ConfigureNHibernate.cs:
```diff
@@ -20,57 +19,62 @@ namespace NServiceBus.Persistence.NHibernate
+            get
+            {
+                string connString;
+                if (!configuration.Properties.TryGetValue(Environment.ConnectionString, out connString))
```
So you are parsing it every time I access this property. Why?

danielmarbach commented on src/NServiceBus.NHibernate/SharedSession/NHibernateStorageSession.cs:
```diff
@@ -23,54 +21,32 @@ internal NHibernateStorageSession()
             context.Container
                 .ConfigureProperty<DbConnectionProvider>(p => p.DisableConnectionSharing, disableConnectionSharing)
-                .ConfigureProperty<DbConnectionProvider>(p => p.DefaultConnectionString, connString);
+                .ConfigureProperty<DbConnectionProvider>(p => p.DefaultConnectionString, configure.ConnectionString);
```
`configure.ConnectionString` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/SharedSession/NHibernateStorageSession.cs:
```diff
@@ -23,54 +21,32 @@ internal NHibernateStorageSession()
 
             context.Container.ConfigureComponent<OpenSqlConnectionBehavior>(DependencyLifecycle.InstancePerCall)
-                .ConfigureProperty(p => p.ConnectionString, connString)
+                .ConfigureProperty(p => p.ConnectionString, configure.ConnectionString)
```
`configure.ConnectionString` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/SharedSession/NHibernateStorageSession.cs:
```diff
@@ -23,54 +21,32 @@ internal NHibernateStorageSession()
 
             context.Container.ConfigureComponent<OpenSessionBehavior>(DependencyLifecycle.InstancePerCall)
-                .ConfigureProperty(p => p.ConnectionString, connString)
+                .ConfigureProperty(p => p.ConnectionString, configure.ConnectionString)
```
`configure.ConnectionString` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/SharedSession/NHibernateStorageSession.cs:
```diff
@@ -23,54 +21,32 @@ internal NHibernateStorageSession()
+            context.Container.ConfigureComponent(b => new NHibernateStorageContext(b.Build<PipelineExecutor>(), configure.ConnectionString), DependencyLifecycle.InstancePerUnitOfWork);
             context.Container.ConfigureComponent<SharedConnectionStorageSessionProvider>(DependencyLifecycle.SingleInstance)
-                .ConfigureProperty(p => p.ConnectionString, connString);
+                .ConfigureProperty(p => p.ConnectionString, configure.ConnectionString);
```
`configure.ConnectionString` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/SharedSession/NHibernateStorageSession.cs:
```diff
@@ -78,7 +54,7 @@ protected override void Setup(FeatureConfigurationContext context)
 
             context.Container.ConfigureComponent<Installer>(DependencyLifecycle.SingleInstance)
-                .ConfigureProperty(x => x.Configuration, configuration)
+                .ConfigureProperty(x => x.Configuration, configure.Configuration)
```
`configure.ConnectionString` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/Subscriptions/NHibernateSubscriptionStorage.cs:
```diff
@@ -24,23 +23,15 @@ public NHibernateSubscriptionStorage()
 
             context.Container.ConfigureComponent<Unicast.Subscriptions.NHibernate.Installer.Installer>(DependencyLifecycle.SingleInstance)
-                .ConfigureProperty(x => x.Configuration, configuration)
+                .ConfigureProperty(x => x.Configuration, configure.Configuration)
```
`configure.Configuration` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/TimeoutPersisters/NHibernateTimeoutStorage.cs:
```diff
@@ -23,43 +22,23 @@ public NHibernateTimeoutStorage()
-                .ConfigureProperty(p => p.ConnectionString, connString)
-                .ConfigureProperty(p => p.SessionFactory, configuration.BuildSessionFactory())
+                .ConfigureProperty(p => p.ConnectionString, configure.ConnectionString)
+                .ConfigureProperty(p => p.SessionFactory, configure.Configuration.BuildSessionFactory())
```
`configure.Configuration` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs:
```diff
@@ -22,23 +21,15 @@ public NHibernateGatewayDeduplication()
 
             context.Container.ConfigureComponent<Deduplication.NHibernate.Installer.Installer>(DependencyLifecycle.SingleInstance)
-                .ConfigureProperty(x => x.Configuration, configuration)
+                .ConfigureProperty(x => x.Configuration, configure.Configuration)
```
`configure.Configuration` bind it to a local variable?

danielmarbach commented on src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs:
```diff
@@ -22,23 +21,15 @@ public NHibernateGatewayDeduplication()
 
             context.Container.ConfigureComponent<Deduplication.NHibernate.GatewayDeduplication>(DependencyLifecycle.SingleInstance)
-                .ConfigureProperty(p => p.SessionFactory, configuration.BuildSessionFactory());
+                .ConfigureProperty(p => p.SessionFactory, configure.Configuration.BuildSessionFactory());
```
`configure.Configuration` bind it to a local variable?

SzymonPobiega commented:
Tries to address #110

SzymonPobiega commented:
@danielmarbach I've merged the two test classes and renamed the methods. I also changed the `ConfigureNHibernate` class into a kind of builder that returns a pair of `Configuration` and a connection string to resolve the issue of re-calculating stuff.

danielmarbach commented:
Looks much cleaner now!

SzymonPobiega commented:
@danielmarbach do you think it is good to merge? I totally forgot about it.

danielmarbach commented:
Based on my last comment I'd say yes :D
